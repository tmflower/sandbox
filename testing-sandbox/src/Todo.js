import axios from "axios";
import { useState } from "react";

export function Todo({ todo }) {
    const [message, setMessage] = useState(null);

    async function selectTask() {
        await axios.get(`http://localhost:3001/todos/${todo.todo_id}`);
    }

    async function deleteTask() {
        await axios.delete(`http://localhost:3001/todos/${todo.todo_id}`);
        setMessage("Task deleted")
    }

    return (
        <div>
            <p onClick={selectTask}><span onClick={deleteTask}>üóëÔ∏è</span> {todo.task}</p>
            {message ? <p>{message}</p> : null}
        </div>
    )
}