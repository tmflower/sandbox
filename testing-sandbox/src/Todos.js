import { useState } from "react";
import axios from "axios";
import { Todo } from "./Todo";

export function Todos({todos}) {

    const [message, setMessage] = useState('');
    const [formData, setFormData] = useState({ task: ''});
    const { task } = formData;

    const handleChange = (evt) => {
        const { name, value } = evt.target;
        setFormData(formData => ({ ...formData, [name]: value}));
    };

    const handleSubmit = async (evt) => {
        evt.preventDefault();
        try {
            await axios.post(`http://localhost:3001/todo`, { task });
            setMessage("Task added!");
            setFormData({task: ''})
        }
        catch(err) {
            console.error(err);
        }
    };

    return (
        <div>
            <h1>Todos</h1>
            <h4>Add a new task</h4>
            <form>
                <label htmlFor="task">Task:
                <input 
                    type="text" 
                    name="task" 
                    value={task} 
                    id="task" 
                    onChange={handleChange}>
                </input>
                </label>
                <button onClick={handleSubmit}>Submit</button> 
            </form>
            <p>{message}</p>
            {!todos.length ? <p>No tasks yet. Add a new task above.</p> 
            :
            <ul>
                {todos.map((todo, i) => <li key={i} style={{listStyleType:'none'}}><Todo key={i} todo={todo}/></li>)}
            </ul>
            }
        </div>
    )
}