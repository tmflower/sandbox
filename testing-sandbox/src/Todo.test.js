import { render, screen } from "@testing-library/react";
import { Todo } from "./Todo";
import { Todos } from "./Todos";
import userEvent from "@testing-library/user-event";

jest.mock("axios");

test("renders without crashing", function() {
    render(<Todo todo={{ task: "Walk the dog" }}/>);
});

test("matches snapshot", () => {
    const { asFragment } = render(<Todo todo={{ task: "Walk the dog"}}/>);
    expect(asFragment()).toMatchSnapshot();
  });

test("renders Todo component correctly", () => {
    render(<Todo todo={{ task: "Walk the dog" }}/>);
    const p = screen.queryByText("Walk the dog");
    expect(p).toBeInTheDocument();
    const deleteIcon = screen.getByText("üóëÔ∏è");
    expect(deleteIcon).toBeInTheDocument();
});

test("handles click to delete", async () => {
    render(<Todo todo={{ task: "Walk the dog", todo_id: 4 }}/>);
    const deleteIcon = screen.getByText("üóëÔ∏è");
    const message = screen.queryByText("Task deleted");
    expect(message).not.toBeInTheDocument();

    await userEvent.click(deleteIcon);

    const messageAfterClick = screen.getByText("Task deleted");
    expect(messageAfterClick).toBeInTheDocument();
})