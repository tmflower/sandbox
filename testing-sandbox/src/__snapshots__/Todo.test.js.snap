// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <p>
      <span>
        🗑️
      </span>
       Walk the dog
    </p>
  </div>
</DocumentFragment>
`;
